[16.0] [ADD] survey_certification_branding: support custom company name and logo in certification report

rrebollo opened the pull request:
# Survey Certification Branding

## Overview
This module extends Odoo's Survey functionality to enable customization of certification reports by allowing:
- Custom company name per certification
- Custom logo per certification

## Features
- Optionally override default company name displayed on certificates
- Optionally set unique logo for each certification (replaces company logo)

![Certification Branding Section in Survey's Form View](https://github.com/user-attachments/assets/a894307a-87ba-48d8-90db-b25cea2db12d)

## Roadmap
- [x] Implement custom fields in survey model
- [x] Add UI controls in survey views
- [x] Update QWeb report templates
- [x] Spanish translation
- [x] Add tests (including certification report rendering)
- [x] Document usage

## Dependencies
- survey (Odoo base module)

@BinhexTeam

christian-ramos-tecnativa commented on survey_certification_branding/report/survey_report_templates.xml:
```diff
@@ -0,0 +1,113 @@
+        </template>
+
+
+        <template
+        id="certification_report_view"
+        inherit_id="survey.certification_report_view"
+    >
+            <xpath expr="//t[@t-call='{{layout_template}}']" positions="inside">
+                <t
+                t-set="is_a_certification"
+                t-value="user_input.survey_id.certification"
+            />
+                <t
+                t-if="is_a_certification"
+                t-set="certification_company_name"
+                t-value="user_input.survey_id.certification_company_name"
+            />
+                <t
+                t-if="is_a_certification"
+                t-set="certification_logo"
+                t-value="user_input.survey_id.certification_logo_512"
+            />
+            </xpath>
+        </template>
```
With the changes above we dont need this
```suggestion
```

christian-ramos-tecnativa commented on survey_certification_branding/report/survey_report_templates.xml:
```diff
@@ -0,0 +1,113 @@
+        inherit_id="survey.certification_report_view_modern"
+        priority="109"
+    >
+            <xpath
+            expr="//span[@t-field='user_input.env.company.display_name']"
+            position="attributes"
+        >
+                <attribute name="t-if">not certification_company_name</attribute>
+            </xpath>
+            <xpath
+            expr="//span[@t-field='user_input.env.company.display_name']"
+            position="after"
+        >
+                <t t-if="certification_company_name">
+                    <span
+                    class="certification-company"
+                    t-out="certification_company_name"
+                />
+                </t>
+            </xpath>
+
+            <xpath
+            expr="//span[@t-field='user_input.env.company.logo']"
+            position="attributes"
+        >
+                <attribute name="t-if">not certification_company_name</attribute>
+            </xpath>
+            <xpath
+            expr="//span[@t-field='user_input.env.company.logo']"
+            position="after"
+        >
+                <t t-if="certification_logo">
+                    <span
+                    class="certification-company-logo"
+                    t-out="certification_logo"
+                    t-options="{'widget': 'image'}"
+                    role="img"
+                />
+                </t>
+            </xpath>
```
```suggestion
            <xpath
            expr="//span[@t-field='user_input.env.company.display_name']"
            position="attributes"
        >
                <attribute name="t-if">not user_input.survey_id.certification_company_name</attribute>
            </xpath>
            <xpath
            expr="//span[@t-field='user_input.env.company.display_name']"
            position="after"
        >
                <t t-else="">
                    <span
                    class="certification-company"
                    t-field="user_input.survey_id.certification_company_name"
                />
                </t>
            </xpath>

            <xpath
            expr="//span[@t-field='user_input.env.company.logo']"
            position="attributes"
        >
                <attribute name="t-if">not user_input.survey_id.certification_logo_512</attribute>
            </xpath>
            <xpath
            expr="//span[@t-field='user_input.env.company.logo']"
            position="after"
        >
                <t t-else="">
                    <span
                    class="certification-company-logo"
                    t-field="user_input.survey_id.certification_logo_512"
                    t-options="{'widget': 'image'}"
                    role="img"
                />
                </t>
            </xpath>
```

christian-ramos-tecnativa commented on survey_certification_branding/report/survey_report_templates.xml:
```diff
@@ -0,0 +1,113 @@
+        inherit_id="survey.certification_report_view_classic"
+        priority="109"
+    >
+            <xpath
+            expr="//span[@t-field='user_input.env.company.display_name']"
+            position="attributes"
+        >
+                <attribute name="t-if">not certification_company_name</attribute>
+            </xpath>
+            <xpath
+            expr="//span[@t-field='user_input.env.company.display_name']"
+            position="after"
+        >
+                <t t-if="certification_company_name">
+                    <span
+                    class="certification-company"
+                    t-out="certification_company_name"
+                />
+                </t>
+            </xpath>
+
+            <xpath
+            expr="//span[@t-field='user_input.env.company.logo']"
+            position="attributes"
+        >
+                <attribute name="t-if">not certification_company_name</attribute>
+            </xpath>
+            <xpath
+            expr="//span[@t-field='user_input.env.company.logo']"
+            position="after"
+        >
+                <t t-if="certification_logo">
+                    <span
+                    class="certification-company-logo"
+                    t-out="certification_logo"
+                    t-options="{'widget': 'image'}"
+                    role="img"
+                />
+                </t>
+            </xpath>
```
```suggestion
            <xpath
            expr="//span[@t-field='user_input.env.company.display_name']"
            position="attributes"
        >
                <attribute name="t-if">not user_input.survey_id.certification_company_name</attribute>
            </xpath>
            <xpath
            expr="//span[@t-field='user_input.env.company.display_name']"
            position="after"
        >
                <t t-else="">
                    <span
                    class="certification-company"
                    t-field="user_input.survey_id.certification_company_name"
                />
                </t>
            </xpath>

            <xpath
            expr="//span[@t-field='user_input.env.company.logo']"
            position="attributes"
        >
                <attribute name="t-if">not user_input.survey_id.certification_logo_512</attribute>
            </xpath>
            <xpath
            expr="//span[@t-field='user_input.env.company.logo']"
            position="after"
        >
                <t t-else="">
                    <span
                    class="certification-company-logo"
                    t-field="user_input.survey_id.certification_logo_512"
                    t-options="{'widget': 'image'}"
                    role="img"
                />
                </t>
            </xpath>
```

Christian-RB reviewed (changes requested):
I proposed a couple of changes to make it simpler, let me know what you think @rrebollo, it's untested code.

rrebollo commented:
@Christian-RB I've implemented your suggestions with some adaptations. Here are the details:

**Changes Made:**
1. Modified the `write` method in `survey` to enforce data integrity:
   - Now ensures branding fields (`certification_company_name`/`certification_logo`) are cleared when a survey is no longer marked as certification
   - Prevents orphaned branding data when certification is unchecked

**Background Rationale:**
There was a potential edge case where:
- Survey gets marked as certification + branding fields populated
- Later gets unchecked as certification
- Branding fields would remain populated (while template conditionals only check certification flag)

what do you think about it?

christian-ramos-tecnativa commented:
I think is a good idea clearing that data, but I would do it with an api.depends and setting the field as `store=True` `readonly=False`. I usually extend the `write` function when I need to compare the new values with the old ones. Ping me when the tests are done to review again. Thanks!

christian-ramos-tecnativa commented:
Also remove the [WIP] tag from the title when finished, please

rrebollo commented:
Hey @Christian-RB, just wanted to let you know everything's ready on my end. I've done some basic testing, I have considered how to properly test the QWeb report rendering - that's not really my strong suit. What do you think?

christian-ramos-tecnativa commented on survey_certification_branding/models/survey_survey.py:
```diff
@@ -0,0 +1,48 @@
+        readonly=False,
+    )
+
+    @api.depends("certification")
+    def _compute_certification_company_name(self):
+        """Clear the certification company name if the survey is not a certification"""
+        self.filtered(
+            lambda survey: not survey.certification
+            and survey.certification_company_name
+        ).update({"certification_company_name": False})
+
+    @api.depends("certification")
+    def _compute_certification_logo_512(self):
+        """Clear the certification logo if the survey is not a certification"""
+        self.filtered(
+            lambda survey: not survey.certification and survey.certification_logo_512
+        ).update({"certification_logo_512": False})
```
IMO you dont need 2 functions to clear the data, as you are checking just the certification field you could do it in one

edescalona reviewed (approved):
Hi @rrebollo , could you please check the runboat?

Tested locally and excellent work. Thanks for your contribution.

dreispt commented:
/ocabot merge nobump
